invenio_rdm_records/vocabularies/__init__.py

```diff
@@ -8,190 +8,111 @@
-        adapt it to their needs easily).
-
-        TODO: Be attentive to generalization for all vocabularies.
-        """
-        options = {'type': [], 'subtype': []}
-
-        for (_type, subtype), entry in self.data.items():
-            type_option = {
-                'icon': entry.get('type_icon'),
-                'text': _(entry.get('type_name')),
-                'value': _type
-            }
-
-            if type_option not in options['type']:
-                options['type'].append(type_option)
-
-            # NOTE: There isn't always a subtype
-            if subtype:
-                subtype_option = {
-                    'parent-text': type_option['text'],
-                    'parent-value': type_option['value'],
-                    'text': _(entry.get('subtype_name')),
-                    'value': subtype,
-                }
-
-                # These are not duplicated so we can just append
-                options['subtype'].append(subtype_option)
-
-        return options
-
-
-class Vocabulary(object):
+class Vocabularies(object):
     """Interface to vocabulary data."""
 
     this_dir = dirname(__file__)
     vocabularies = {
-        # NOTE: keys should be same as MetadataSchemaV1 fields
+        # NOTE: dotted keys should parallel MetadataSchemaV1 fields
         'resource_type': {
             'path': join(this_dir, 'resource_types.csv'),
             'class': ResourceTypeVocabulary,
             'object': None
+        },
+        'contributors.role': {
+            'path': join(this_dir, 'contributor_role.csv'),
+            'class': ContributorRoleVocabulary,
+            'object': None
+        },
+        'titles.type': {
+            'path': join(this_dir, 'title_type.csv'),
+            'class': TitleTypeVocabulary,
+            'object': None
+        },
+        'access_right': {
+            'path': join(this_dir, 'access_right.csv'),
+            'class': AccessRightVocabulary,
+            'object': None
         }
+
     }
 
     @classmethod
-    def get_vocabulary(cls, vocabulary_type):
-        """Returns the corresponding Vocabulary object."""
-        obj = cls.vocabularies.get(vocabulary_type, {}).get('object')
+    def get_vocabulary(cls, key):
+        """Returns the Vocabulary object corresponding to the path.
+
+        :param key: string of dotted subkeys for Vocabulary object.
+        """
+        vocabulary_dict = cls.vocabularies.get(key)
+
+        if not vocabulary_dict:
+            return None
+
+        obj = vocabulary_dict.get('object')
+
         if not obj:
-            path = (
+            custom_vocabulary_dict = (
                 current_app.config
                 .get('RDM_RECORDS_CUSTOM_VOCABULARIES', {})
-                .get(vocabulary_type) or
-                cls.vocabularies
-                .get(vocabulary_type)
-                .get('path')
+                .get(key, {})
+            )
+
+            path = (
```

Comment:
understanding check: we do this so defaults are the above `cls.vocabularies` but we five preference to the configured ones? If we are gonna get from config anyway, wouldn't it be good to have the defaults defined in config (so they are only in one place)?

Reply:
The developer experience this allows is more convenient IMHO:

As here:

If the developer wants to override the contributors role, they just have to do:

```python
# invenio.cfg

RDM_RECORDS_CUSTOM_VOCABULARIES = {
    'contributors.role': {
        'path': '/path/to/my/contributor_role.csv'
    }
}
```

Otherwise, if we were to use `RDM_RECORDS_CUSTOM_VOCABULARIES` for all vocabularies:

If the developer wants to override the contributors role, they have to do:

```python
# invenio.cfg
from invenio_rdm_records.config import RDM_RECORDS_CUSTOM_VOCABULARIES

RDM_RECORDS_CUSTOM_VOCABULARIES.update({
    'contributors.role': {
        'path': '/path/to/my/contributor_role.csv'
    }
})
```

It's not a big difference, but it's the little things that matter ;) .

We also store the loaded Vocabulary instances in the Vocabularies' `object` field which should not be exposed to others.

Reply:
Makes sense :)